src/sharing/index.js

```diff
@@ -15,11 +47,260 @@ export const getPrimaryCozy = contact =>
+
+export default class SharingProvider extends Component {
+  constructor(props, context) {
+    super(props, context)
+    this.state = {
+      byDocId: {},
+      sharings: [],
+      permissions: [],
+      documentType: props.documentType || 'Document',
+      isOwner: docId => isOwner(this.state, docId),
+      getOwner: docId => getOwner(this.state, docId),
+      getSharingType: docId => getSharingType(this.state, docId),
+      getRecipients: docId => getRecipients(this.state, docId),
+      getSharingLink: document => getSharingLink(this.state, document),
+      share: this.share,
+      revoke: this.revoke,
+      revokeSelf: this.revokeSelf,
+      shareByLink: this.shareByLink,
+      revokeSharingLink: this.revokeSharingLink
+    }
+  }
+
+  dispatch = action =>
+    this.setState(state => ({ ...state, ...reducer(state, action) }))
+
+  componentDidMount() {
+    Promise.all([
+      this.context.client
+        .collection('io.cozy.sharings')
+        .findByDoctype(this.props.doctype),
+      this.context.client
+        .collection('io.cozy.permissions')
+        .findLinksByDoctype(this.props.doctype),
+      this.context.client.collection('io.cozy.permissions').findApps()
+    ]).then(([sharings, permissions, apps]) =>
+      this.dispatch(
+        receiveSharings({
+          sharings: sharings.data,
+          permissions: permissions.data,
+          apps: apps.data
+        })
+      )
+    )
+  }
+
+  share = async (document, recipients, sharingType, description) => {
+    const resp = await this.context.client
+      .collection('io.cozy.sharings')
+      .share(document, recipients, sharingType, description)
+    this.dispatch(addSharing(resp.data))
+    return resp.data
+  }
+
+  revoke = async (document, recipientEmail) => {
+    const sharing = getSharingForRecipient(
+      this.state,
+      document.id,
+      recipientEmail
+    )
+    await this.context.client
+      .collection('io.cozy.sharings')
+      .revokeRecipient(sharing, recipientEmail)
+    this.dispatch(revokeRecipient(sharing, recipientEmail))
+  }
+
+  revokeSelf = async document => {
+    const sharing = getSharingForSelf(this.state, document.id)
+    await this.context.client.collection('io.cozy.sharings').revokeSelf(sharing)
+    this.dispatch(revokeSelf(sharing))
+  }
+
+  shareByLink = async document => {
+    trackSharingByLink(document)
+    const resp = await this.context.client
+      .collection('io.cozy.permissions')
+      .createSharingLink(document)
+    this.dispatch(addSharingLink(resp.data))
+    return resp
+  }
+
+  revokeSharingLink = document => {
+    // Because some duplicate links have been created in the past, we must ensure
+    // we revoke all of them
+    const perms = getDocumentPermissions(this.state, document.id)
+    return Promise.all(
+      perms.map(p =>
+        this.context.client.collection('io.cozy.permissions').destroy(p)
+      )
+    ).then(() => this.dispatch(revokeSharingLink(perms)))
```

Comment:
Can we switch this to `async` / `await`? It'd be easier to read, and the rest of the file seems to be using `async`.